can now set value prop to parent's state and an option to only notify on user input changes

MatthewMaclean opened the pull request:
This change allows the input box's value to be changed via it's props rather than only taking an initial value. It also adds an option to only notify on user input changes and not due to the props changes to allow for creating a history of user inputs allowing for undo and redo.

nkbt commented:
Welcome, @MatthewMaclean!

I've looked at your changes, everything seems fine. I have a question about the concept itself (usecase). It would be awesome if you can give some simple example (in addition to "feature-test" in App) that shows this feature within some context. I mean - something that actually uses this feature.

I am also thinking if it is backward compatible and it might not be since we now check `this.props.value` and not `this.state.value` (thus, behaviour is changed). There is nothing wrong with it, just want to make sure major version is needed and not minor after merge

MatthewMaclean commented:
I added a history feature to the App so that you can see the use-case and removed the "set to ASDF" button as it wasn't required to show the feature anymore.

I think it is a breaking change and the major version should be incremented.
There are three considerations for backwards compatibility:
-Value was used for initial value-
This would not break backwards compatibility as it only occurred at the beginning.

-Value was stored in state of parent element-
This would only be possible if you wrapped the input in a component that had the check I switched it to. In which case, behaviour should remain the same.

-Value was provided as a variable that can change and replace current contents (not state mirroring value)-
This would have different behaviour. I don't have a use case for it, but it is possible.

nkbt commented:
I found one issue with example. When you hit Return it duplicates items in history. Should be only one added (last)

![enter-press](https://cloud.githubusercontent.com/assets/175264/11713766/7ac0689c-9f89-11e5-82b1-6e58fcc2fc17.gif)

nkbt commented:
I found it is possible to do everything without adding extra flag (simplified DebounceInput code a bit, so when it is Controllable - it works well)

![undo-redo](https://cloud.githubusercontent.com/assets/175264/11715207/1b70d65a-9f94-11e5-9bb0-a06c96186565.gif)


See #13

nkbt commented:
I merged it to have Undo/Redo example. But I will remove extra flag. I don't think it is necessary

nkbt commented:
Thanks a lot

MatthewMaclean commented:
Passing through the event is great too. When do you think you'll cut a release?

nkbt commented:
I am deprecating some things in react-component-template... then update this one accordingly. I reckon today can be a day for 2.0.0 :)

nkbt commented:
@MatthewMaclean Published 2.0.1

See changes:
- https://github.com/nkbt/react-debounce-input/releases/tag/v2.0.0
- https://github.com/nkbt/react-debounce-input/releases/tag/v2.0.1
